### ISSUE TITLE
./coap-server segfault 2

### ISSUE THREAD

**Issue by 16201535wangshenyu** [obgm/libcoap#971]
## Environment
   
- libcoap version (run ``git describe --tags`` to find it):
- 
    v4.3.0-1da37b9

- Build System:             Make
- Operating System:         Linux 
- Operating System Version: [5.4.0-124-generic #140~18.04.1-Ubuntu SMP Fri Aug 5 11:43:34 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux ]
- Hosted Environment:       [None]

## Problem Description
   I write a fuzzing tool to fuzz the coap-server. when I send a udp frame to the coap-server , the program(coap-server) is goes crashed. the coap-server used the default configure.

### Expected Behavior

 The program is noramlly running.

### Actual Behavior

  the coap-server  crashed with blow infomation:

### Steps to reproduce

1.  bulid a udp client 
2. send the blow coap frame
```c
 No.3454 TX :5006da6e432e63402101121633456173796e63113c01ff41751170a2c5b1c4636f6170d1ce10ff0d680371ec9e9ac63739fe
 Fatal error! ########################################## 
 Exception: timed out
 CORE DUMP: 15.37.08#2022-11-18
```

### GDB core_dum.file
  
![image](https://user-images.githubusercontent.com/51477879/202696568-e4f0b1ae-9e4e-49af-b472-9a0290fbae56.png)

## Debug Logs

段错误 (核心已转储)

## Other items if possible

- [ ] Does what you are trying to do work under any configuration.  Detail what
works.
 I try to fix the bug, therefore, The code i changed a liitle. 
 I would like to provide the coap-server executable as well as the core-dump file if possible.
 contact me，the core dump file is too large ，i can‘t upload.
 the executable file:
https://github.com/16201535wangshenyu/coap_fuzz/blob/master/core3/coap-server.zip
  my email:51205902099@stu.ecnu.edu.cn
- [ ] Network configuration
       coap default configure. jun run ./coap_server
- [ ] The project Build 
       ```shell
           ./autogen.sh 
          ./configure --enable-examples --enable-tests --enable-gcov  --disable-doxygen --disable-manpages --disable-dtls --prefix=~ /tools/libcoap-4.3.0/install
          make 
          make install 
       ```
### The change of the project。
     all the place I changed ,i used a macro BUG_FIX
      src/resource.c 
   ```c
#ifndef BUG_FIX
#define BUG_FIX
#endif

static void
coap_notify_observers(coap_context_t *context, coap_resource_t *r,
                      coap_deleting_resource_t deleting) {
  coap_method_handler_t h;
  coap_subscription_t *obs;
  coap_binary_t token;
  coap_pdu_t *response;
  uint8_t buf[4];
  coap_string_t *query;
  coap_block_t block;

  if (r->observable && (r->dirty || r->partiallydirty)) {
    r->partiallydirty = 0;

    LL_FOREACH(r->subscribers, obs) {
      if (r->dirty == 0 && obs->dirty == 0) {
        /*
         * running this resource due to partiallydirty, but this observation's
         * notification was already enqueued
         */
        context->observe_pending = 1;
        continue;
      }
      if (obs->session->con_active >= COAP_DEFAULT_NSTART &&
          ((r->flags & COAP_RESOURCE_FLAGS_NOTIFY_CON) ||
           (obs->non_cnt >= COAP_OBS_MAX_NON))) {
        r->partiallydirty = 1;
        obs->dirty = 1;
        context->observe_pending = 1;
        continue;
      }

      coap_mid_t mid = COAP_INVALID_MID;
      obs->dirty = 0;
      /* initialize response */
      response = coap_pdu_init(COAP_MESSAGE_CON, 0, 0, coap_session_max_pdu_size(obs->session));
      if (!response) {
        obs->dirty = 1;
        r->partiallydirty = 1;
        context->observe_pending = 1;
        coap_log(LOG_DEBUG,
                 "coap_check_notify: pdu init failed, resource stays "
                 "partially dirty\n");
        continue;
      }
      #ifdef BUG_FIX
        if (!obs->pdu ||(long long)(obs->pdu) < 0x550000000000 || !coap_add_token(response, obs->pdu->token_length, obs->pdu->token)) {
                         

      #else
        if (!coap_add_token(response, obs->pdu->token_length, obs->pdu->token)) {
            // return 0;

      #endif

    //   if (!coap_add_token(response, obs->pdu->token_length, obs->pdu->token)) {
        obs->dirty = 1;
        r->partiallydirty = 1;
        context->observe_pending = 1;
        coap_log(LOG_DEBUG,
                 "coap_check_notify: cannot add token, resource stays "
                 "partially dirty\n");
        coap_delete_pdu(response);
        continue;
      }

      token.length = obs->pdu->token_length;
      token.s = obs->pdu->token;

      obs->pdu->mid = response->mid = coap_new_message_id(obs->session);
      if ((r->flags & COAP_RESOURCE_FLAGS_NOTIFY_CON) == 0 &&
          ((r->flags & COAP_RESOURCE_FLAGS_NOTIFY_NON_ALWAYS) ||
           obs->non_cnt < COAP_OBS_MAX_NON)) {
        response->type = COAP_MESSAGE_NON;
      } else {
        response->type = COAP_MESSAGE_CON;
      }
      switch (deleting) {
      case COAP_NOT_DELETING_RESOURCE:
        /* fill with observer-specific data */
        coap_add_option(response, COAP_OPTION_OBSERVE,
                        coap_encode_var_safe(buf, sizeof (buf),
                                             r->observe),
                        buf);
        if (coap_get_block(obs->pdu, COAP_OPTION_BLOCK2, &block)) {
          /* Will get updated later (e.g. M bit) if appropriate */
          coap_add_option(response, COAP_OPTION_BLOCK2,
                          coap_encode_var_safe(buf, sizeof(buf),
                                               ((0 << 4) |
                                                (0 << 3) |
                                                block.szx)),
                          buf);
        }

        h = r->handler[obs->pdu->code - 1];
        assert(h);      /* we do not allow subscriptions if no
                         * GET/FETCH handler is defined */
        query = coap_get_query(obs->pdu);
        h(r, obs->session, obs->pdu, query, response);
        /* Check if lg_xmit generated and update PDU code if so */
        coap_check_code_lg_xmit(obs->session, response, r, query);
        coap_delete_string(query);
        if (COAP_RESPONSE_CLASS(response->code) != 2) {
          coap_remove_option(response, COAP_OPTION_OBSERVE);
        }
        if (COAP_RESPONSE_CLASS(response->code) > 2) {
          coap_delete_observer(r, obs->session, &token);
        }
        break;
      case COAP_DELETING_RESOURCE:
      default:
        response->type = COAP_MESSAGE_NON;
        response->code = COAP_RESPONSE_CODE(404);
        break;
      }

      if (response->type == COAP_MESSAGE_CON ||
          (r->flags & COAP_RESOURCE_FLAGS_NOTIFY_NON_ALWAYS)) {
        obs->non_cnt = 0;
      } else {
        obs->non_cnt++;
      }

      mid = coap_send_internal( obs->session, response );

      if (COAP_INVALID_MID == mid) {
        coap_log(LOG_DEBUG,
                 "coap_check_notify: sending failed, resource stays "
                 "partially dirty\n");
        obs->dirty = 1;
        r->partiallydirty = 1;
        context->observe_pending = 1;
      }

    }
  }
  r->dirty = 0;
}

 ```
src/pdu.c
```c
#ifndef BUG_FIX
#define BUG_FIX
#endif
int
coap_add_token(coap_pdu_t *pdu, size_t len, const uint8_t *data) {
  /* must allow for pdu == NULL as callers may rely on this */
  #ifdef BUG_FIX
        if (!pdu ||(long long)pdu < 0x550000000000 || len > 8)
            return 0;                   

  #else
        if (!pdu || len > 8)
            return 0;

  #endif

  if (pdu->used_size) {
    coap_log(LOG_WARNING,
             "coap_add_token: The token must defined first. Token ignored\n");
    return 0;
  }
  if (!coap_pdu_check_resize(pdu, len))
    return 0;
  pdu->token_length = (uint8_t)len;
  if (len)
    memcpy(pdu->token, data, len);
  pdu->max_opt = 0;
  pdu->used_size = len;
  pdu->data = NULL;

  return 1;
}
```

**Comment by mrdeep1** [comment]
Running this with `1da37b9`  (same as release 4.3.0) I get
````
Nov 18 14:36:26.171 DEBG created UDP  endpoint [::]:5683
Nov 18 14:36:26.171 DEBG created TCP  endpoint [::]:5683
Nov 18 14:36:37.291 DEBG ***[::ffff:127.0.0.1]:5683 <-> [::ffff:127.0.0.1]:42758 (if1) UDP : new incoming session
Nov 18 14:36:37.291 DEBG *  [::ffff:127.0.0.1]:5683 <-> [::ffff:127.0.0.1]:42758 (if1) UDP : received 50 bytes
v:1 t:NON c:PATCH i:da6e {} [ ETag:0x2e6340, Observe:1, Uri-Port:5683, Uri-Path:async, Content-Format:application/cbor, Content-Format:255, Hop-Limit:117, Accept:application/senml+cbor, Block1:3163/_/32, Proxy-Scheme:coap, No-Response:\x10 ] :: binary data length 11
<<0d680371ec9e9ac63739fe>>
<<. h . q . . . . 7 9 . >>
Nov 18 14:36:37.292 DEBG Proxy-Scheme support not configured
Nov 18 14:36:37.292 DEBG *  [::ffff:127.0.0.1]:5683 <-> [::ffff:127.0.0.1]:42758 (if1) UDP : sent 27 bytes
v:1 t:NON c:5.05 i:da6e {} [ ] :: 'Proxying Not Supported'
````
Now, as the core file shows that it crashed having been called from `coap_notify_observers() `, this would have been triggered from a previous packet setting up the observe, not the current packet you have provided.  Again as in #970, #874 may fix your issue (#874 is in release 4.3.1).  Can you please confirm if #874 fixes your issue?

**Comment by mrdeep1** [comment]
Thanks for the testing #874 and feedback.  Closing this Issue as fixed.
